syntax = "proto3";

package notification_hub;

option go_package = "pb";

// Notification Task configuration
message NotificationTask {
    // A unique identifier for the notification
    string ID = 1;
    // name of the task
    string name = 2;
    // The Message type. Choices are: Transactional/Marketing (defaults to Marketing)
    enum MessageType {
        MARKETING = 0;
        TRANSACTIONAL = 1;
    }
    MessageType messageType = 3;
    // channel specific details
    Email email = 4;
    SMS sms = 5;
    Whatsapp whatsapp  = 6;
    Push push = 7;
    // SentByID
    string sentByID = 8;
    // Client
    string client = 9;
    // Waterfall mode : Override / Auto
    enum WaterfallMode {
        AUTO = 0;
        OVERRIDE = 1;
    }
    WaterfallMode waterfallType = 10;
    // timestramp - UNIX EPOCH time
    int64 triggeredOn = 11;
    // platform
    enum Platform {
        AASAANJOBS = 0;
        OLXPEOPLE = 1;
    }
    Platform platform = 12;
}
// Waterfall configuration
message Waterfall {
    // The priority of the channel in the task.
    int32 priority = 1;
    // The time offset in seconds after which the notification should be triggered after the previous channel
    int32 offsetTime = 2;
}
// SMS configuration
message SMS {
    // URL to the notification template (should be stored in any cloud storage)
    string template = 1;
    // The mobile number to whom the sms is intended to be sent to
    string mobile = 2;
    // The template placeholder data
    string context = 3;
    // The waterfall configuration for the channel
    Waterfall waterfallConfig = 4;
    // expiry - timestramp UNIx EPOCH time
    int64 expiry = 5;
    // User ID
    string userID = 6;
}
// Whatsapp configuration
message Whatsapp {
    // URL to the notification template (should be stored in any cloud storage)
    string template = 1;
    // The mobile number to whom the sms is intended to be sent to
    string mobile = 2;
    // The template placeholder data
    string context = 3;
    // The waterfall configuration for the channel
    Waterfall waterfallConfig = 4;
    // expiry - timestramp UNIx EPOCH time
    int64 expiry = 5;
    // User ID
    string userID = 6;
}

message EmailRecipient {
    // The email ID of the recipient
    string email = 1;
    // The name of the person or company that is sending the email
    string name = 2;
    // User ID
    string userID = 3;
}

message EmailAttachment{
    // The filename of the attachment
    string filename = 1;
    // The URL of the file that will be sent as an attachment
    string url = 2;
}

// Email configuration
message Email {
    // URL to the notification template (should be stored in any cloud storage)
    string template = 1;
    // Email Recipient
    // An array of email recipients
    repeated EmailRecipient toRecipients = 2;
    // The template placeholder data
    string context = 3;
    // The subject of your email
    string subject = 4;
    // The sender of the email (should be an email recipient object)
    EmailRecipient sender = 5;
    // The email to which the user can reply to by default (should be an email recipient object)
    EmailRecipient replyTo = 6;
    // An array of recipients who will receive a copy of your email
    repeated EmailRecipient ccRecipients = 7;
    // An array of objects in which you can specify any attachments you want to include
    // Email attachment
    repeated EmailAttachment attachments = 8;
    // The waterfall configuration for the channel
    Waterfall waterfallConfig = 9;
    // expiry - timestramp UNIx EPOCH time
    int64 expiry = 10;
}
// Push
message Push {
    // URL to the notification template (should be stored in any cloud storage)
    string template = 1;
    // An array of Amazon Resource Names, which corresponds to the registration number of an Android device
    string token = 2;
    // The template placeholder data
    string context = 3;
    // The waterfall configuration for the channel
    Waterfall waterfallConfig = 4;
    // expiry - timestramp UNIx EPOCH time
    int64 expiry = 5;
    // User ID
    string userID = 6;

    string extraPayload = 7;

    enum ClientApp {
        ANDROID = 0;
        IOS = 1;
        WEB = 2;
    }
    ClientApp clientApp = 8;
}
